crates/typst/src/layout/inline/mod.rs

```diff
@@ -1446,3 +1478,50 @@ fn overhang(c: char) -> f64 {
         _ => 0.0,
     }
 }
+
+/// Whether the hyphen should repeat at the begin of the next line
+fn should_repeat_hyphen(pred_line: &Line) -> bool {
+    // If the predecessor line does not end with a Dash::HardHyphen, we shall not place a hyphen at
+    // the beginning of the next line.
+    if pred_line.dash != Some(Dash::HardHyphen) {
+        return false;
+    }
+
+    // If there's a trimmed out space, we needn't repeat the hyphen. That's the case of a text like
+    // "... kebab é a -melhor- comida que existe", where the hyphens are a kind of emphasis marker.
+    if pred_line.trimmed.end != pred_line.end {
+        return false;
+    }
+
+    // The hyphen should repeat only in the languages that requires that feature.
+    // For more information see the discussion at https://github.com/typst/typst/issues/3235
+    if let Some(Item::Text(shape)) = pred_line.last.as_ref() {
+        match shape.lang.as_str() {
```

Comment:
Imho we should make this configurable within `TextElem`, something like `repeat-hyphenation-on-linebreak` (maybe a shorter name), which can be either `true`, `false` or `auto` (default, which varies depending on the language).

Reply:
Indeed, a configuration within `TextElem` would make things more flexible. But the rules for hyphen duplication may have some tricky exceptions that are hard to express in configurable attributes --- for example: in Spanish we must not repeat the hyphen if the next word is capitalized.

Reply:
I don't know if there's other exceptions like that in other languages grammars.

Reply:
I think the option shouldn't be that flexible in principle. `true` would correspond to always doing it, while for `auto` we can accurately map the language to the correct behavior.

That way, you can manually change it to `false` for some sections of text as needed, which should allow for a good degree of customization already.

Reply:
> ... something like repeat-hyphenation-on-linebreak (maybe a shorter name)

What about `linebreak-hyphen`?

The possible values would be
- `"single"`: a single hyphen is placed at the end of the line
- `"repeat"`: the hyphen is also repeated at the beginning of the next line
- `auto` (default): the language rules are accurately mapped to the correct behavior.

Reply:
Seems ok to me as well.

Reply:
With that said, it could be a good idea to hold off implementing that parameter before further input from the maintainers. (cc @laurmaedje)

Reply:
In Spanish and Portuguese, at least[^1], the hyphenation is an orthographic rule. So if you do not obey it, you are making a orthographic mistake, such as misspelling a word. If we do not allow to config that feature, those mistakes will be avoided. 

On the other hand, we don't implement the hyphenation exhaustively, maybe in other languages the hyphen duplication is more a matter of style.

[^1]: The fonts I got this are the RAE orthographic guide and the 1990 Portuguese Orthographic Agreement (the current one). I can read in that two languages. For the other languages I am trusting on automatic translation of the fonts.

Reply:
Let's leave it non-configurable then. When other languages come up, we can add support for them as well.

Reply:
As I see it, the main problem is that using `#set text(lang: ...)` comes with several features which one might or might not want. Most of them can be overridden - e.g. supplements. I don't see why that wouldn't be the case here either - one might want to set the lang for reasons other than this, and/or configure hyphenation despite their `lang` setting.

I'd say being able to at least toggle between always repeating hyphenation and never doing so would be nice, in order to allow users to have maximum control over their documents. There might be differing requirements depending not only on the language but also on the region, or the target journal for example.

With that said, that's not something that is strictly necessary for this PR - it could be implemented separately, and could have further discussion first if needed; just sharing my 2 cents.

Reply:
I'm a bit concerned about the proliferation of very specific options, especially on the `text` function which already has a lot of options. Every extra knob incurs some UX complexity (even if users can mostly ignore it). 

I see how this one is perhaps a gray area, but generally the shaping pipeline might do all sorts of language-specific things which are just automatic and this feels similar.

The user perspective aside, I'm also a bit concerned about this from an implementation perspective. A single property obviously doesn't make a difference, but doing tons of style chain lookups for every single paragraph can also get expensive (especially if they end up happening more than once per paragraph; that's why many properties are cached in the `Preparation` struct).

Reply:
I see what you mean.

While I see the value of having this option, I can also see how it's not really needed at this moment, and I think you have a good argument. In this case, we should likely wait until someone requests this feature to have a better idea of potential use cases and design the API around them, if needed.